iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift

```diff
@@ -0,0 +1,304 @@
+//
+//  TimelineLayout.swift
+//  TimelineExperiment
+//
+//  Created by Martin Young on 11/16/21.
+//
+
+import Foundation
+import UIKit
+
+protocol TimelineCollectionViewLayoutDataSource: AnyObject {
+    func getMessage(at indexPath: IndexPath) -> Messageable?
+}
+
+/// A custom layout for conversation messages. Up to two message cell sections are each displayed as a stack along the z axis.
+/// The stacks appear similar to Apple's Time Machine interface, with the newest message in front and older messages going out into the distance.
+/// As the collection view scrolls up and down, the messages move away or toward the user.
+class TimelineCollectionViewLayout: UICollectionViewLayout {
+
+    private typealias SectionIndex = Int
+
+    weak var dataSource: TimelineCollectionViewLayoutDataSource?
+
+    /// The size of the cells.
+    var itemSize = CGSize(width: 200, height: 100)
+
+    /// A cache of item layout attributes so they don't have to be recalculated.
+    private var cellLayoutAttributes: [IndexPath : UICollectionViewLayoutAttributes] = [:]
+    /// A dictionary of z ranges for all the items. A z range represents the range that each item will be frontmost in its section
+    /// and its scale will be unaltered.
+    private var zRangesDict: [IndexPath : Range<CGFloat>] = [:]
+    /// The current position along the Z axis. This is based off of the collectionview's Y content offset.
+    private var zPosition: CGFloat {
+        return self.collectionView?.contentOffset.y ?? 0
+    }
```

Comment:
The items are laid out along the z axis. I'm using the y content offset to simulate moving along the z axis.